platform/search/src/GenericSearchProvider.js

```diff
@@ -0,0 +1,267 @@
+                    //   it's a JSON object.
+                    message = {
+                        request: 'index',
+                        model: domainObject.getModel(),
+                        id: domainObject.getId()
+                    };
+                    worker.postMessage(message);
+                }
+            }
+            
+            // Tell the worker to search for items it has that match this searchInput.
+            // Takes the searchInput, as well as a max number of results (will return 
+            //   less than that if there are fewer matches).
+            function workerSearch(searchInput, maxResults, timestamp, timeout) {
+                var message = {
+                    request: 'search',
+                    input: searchInput,
+                    maxNumber: maxResults,
+                    timestamp: timestamp,
+                    timeout: timeout
+                };
+                worker.postMessage(message);
+            }
+            
+            // Handles responses from the web worker. Namely, the results of 
+            // a search request. 
+            function handleResponse(event) {
+                var ids = [],
+                    id;
+                
+                // If we have the results from a search 
+                if (event.data.request === 'search') {
+                    // Convert the ids given from the web worker into domain objects
+                    for (id in event.data.results) {
+                        ids.push(id);
+                    }
+                    objectService.getObjects(ids).then(function (objects) {
+                        var searchResults = [],
+                            id;
+                        
+                        // Create searchResult objects
+                        for (id in objects) {
+                            searchResults.push({
+                                object: objects[id],
+                                id: id,
+                                score: event.data.results[id]
+                            });
+                        }
+                        
+                        // Resove the promise corresponding to this 
+                        pendingQueries[event.data.timestamp].resolve({
+                            hits: searchResults,
+                            total: event.data.total,
+                            timedOut: event.data.timedOut
+                        });
+                    });
+                }
+            }
+            
+            worker.onmessage = handleResponse;
+            
+            // Helper function for getItems(). Indexes the tree.
+            function indexItems(nodes) {
+                nodes.forEach(function (node) {
+                    var id = node && node.getId && node.getId();
+                    
+                    // If we have not yet indexed this item, index it 
+                    if (!indexed[id]) {
```

Comment:
To prevent a level of indentation and tidy up the code, there's a much clearer shortcut for the "don't index if already indexed" behavior:

```javascript
if (indexed[id]) {
    return;
}
```